{"title": "Python中循环导入和依赖错误 - virusdefender's blog ", "index": "django,python", "content": "最近使用Python + Django，是一个巨复杂的系统，才几天就40多张数据库的表了。因为在Django里面，表都使用类来实现的，所以就要一开始写很多的class，语法都没问题，但是有时候就会出现ImportError Can not import name xxxx\n\n这个问题遇见了两次，稍微有点差别，现在记录一下来。\n\n1.类之间的循环依赖\n代码大致是这样的\n\nclass Employee(models.Model):\n    permission = models.ManyToManyField(Permission)\n\n\nclass Permission(models.Model):\n    dealer = models.ForeiginKey(Dealer)\n    name = models.CharField(max_length=20)\n\n\nclass Dealer(models.Model):\n    employees = models.ManyToManyField(Employee)\n\n\n这样，Employee依赖Permission，Permission依赖Dealer，Dealer又会回到Employee，这样肯定就是错误的了。\n\n2.不同模块之间的\nService.model 模块\n\nfrom Appointment.models import Appointment\n\nclass Order(models.Model):\n    appointment = models.ForeignKey(Appointment)\n\n\n在Appointment.models里面\n\nfrom Service.models import ServiceItem\n\n\n这样也会引用错误，这个原因找了好长时间。\n因为Order在引用Appointment的时候，会回到Service.models 查找Appointment引用的ServiceItem，但是这个时候Order的构建还没有完成，就会导入错误。这个和Python的机制有关系的。\n\n3.解决办法\n今天又遇见这问题了，虽然找到了原因，但是不知道怎么去修改，因为毕竟业务逻辑在这，现在有stackoverflow上找到一个答案，真心好用。\n也就是使用字符串表示模块，而不进行导入了。http://stackoverflow.com/questions/4379042/django-circular-model-import-help\n\nclass Service(models.Model):\n    appointment = models.ForeignKey(\"appointment.Appointment\")\n\n\n参考http://www.douban.com/group/topic/43938606/http://www.oschina.net/question/919901_88601http://www.oschina.net/translate/top-10-mistakes-that-python-programmers-make\n\n                ", "mainLikeNum": ["0 "], "mainBookmarkNum": "3"}